### PR TITLE
macros: Introduce sameType(a, b) for node types

### PR DESCRIPTION
**Pull request by nanoant**
Previously introduced node comparison `==` was working somehow wrong on nodes
returned from getType(a), comparing just ids of the symbols.

Recently introduced `==` change 47dce26
started comparing symbol nodes pointer-wise, thus strictly. Since getType(a)
always creates new symbol pointing to the type, comparing two such nodes using
`==` always returns false, even they point to the same type.

That is why we need a new sameType macro to be able to tell if these nodes
point to the same type.

### CODE REVIEW

File: compiler/vm.nim [hunk 1]
```diff
@@ -692,6 +692,9 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
       regs[ra].intVal =
         ord(exprStructuralEquivalent(regs[rb].node, regs[rc].node,
                                      strictSymEquality=true))
+    of opcEqNodeType:
+      decodeBC(rkInt)
+      regs[ra].intVal = ord(regs[rb].node.typ == regs[rc].node.typ)
```
Comment (Araq) [inline compiler/vm.nim]: Use types.sameType instead.

Comment (nanoant) [inline compiler/vm.nim]: @Araq Hmm, what if I want to check if the types are equal, not the "same". Eg.:

```nim
type
   SomeInt = int

SomeInt.sameType int #-> true, but SomeInt.typ != int.typ
```

Comment (Araq) [inline compiler/vm.nim]: Then you can compare the string representations of getType. That said, I don't think there is a single "identicalType" comparison in the whole compiler, so I don't think there are valid use cases for this.

**Comment by nanoant** [comment]
@Araq Okay, made separate helpers `identicalType` and (slower) `sameType`. There is no `identicalType` internal proc because this comparison is just `.typ == .typ`, but since macros have no direct access to `.typ` that's why I prefer to have `identicalType` as well as `sameType`.

**Comment by Araq** [comment]
You still haven't told me why ``identicalType`` is a useful thing.

**Comment by nanoant** [comment]
@Araq because there is no way to compare two type nodes, as there is no `.typ` accessor for macros, counterpart to regular node `==` operator. Also in some cases you want to know if something that gets passed to macro is real type or alias, which gives you some extra flexibility:

```nim
import macros

type
  TaintedInt = int

macro check(t: typed): expr =
  if TaintedInt.getType.identicalType t.getType:
    return newLit("tainted")
  else:
    return newLit("normal")

echo check(1)
echo check(TaintedInt(1))
```

**Comment by Araq** [comment]
But a macro that distinguishes between "TaintedInt" and "int" when the actual type system in Nim does not sounds like a horrible idea. For TaintedInt we have 'distinct int' in the language.

**Comment by nanoant** [comment]
I do agree, but I think there's some flaw in current design of Nim compiler that `int` and `TaintedInt` point to different `PType`, even, as you say, they represent the same type.

I would expect that `int` and `TaintedInt` are symbols that point to the same single `PType`. Any reason that is does not work like that?

I don't really understand why we need that, and why we need some complicated comparison process to figure out if two types are the same.

**Comment by Araq** [comment]
> I would expect that int and TaintedInt are symbols that point to the same single PType. Any reason that is does not work like that?

Yes. We support ``type FooInt {.importc: "long int".} = int``.

> I don't really understand why we need that, and why we need some complicated comparison process to figure out if two types are the same.

I don't understand what this has to do with the types API that macros.nim exposes.

**Comment by nanoant** [comment]
@Araq Okay, lets stick with `sameType` only and try if it works for us.

**Comment by Araq** [comment]
You also need to update semfold, ccgexprs and perhaps jsgen. And before you complain: Yes it's bad that they neither use exhaustive case statements nor sensible defaults.

**Comment by nanoant** [comment]
@Araq I am not sure if I follow you, why do I need to update semfold, ccgexprs & jsgen again? Is there something that has to be done there when adding new magic?

**Comment by Araq** [comment]
Yes, just look at the big case statements dealing with magics.

**Comment by nanoant** [comment]
@Araq: `mEqNimrodNode` is not handled there, any reason `mSameNodeType` should be? I don't think it is any different.

**Comment by nanoant** [comment]
@Araq Since we agreed that `NimNode` cannot be constant so `semfold` does not apply, also it will be used only in macros so similarly putting it into `jsgen` does not make sense, is there anything that stops us from merging that?

**Comment by dom96** [comment]
Seems this has broken bootstrapping.

**Comment by nanoant** [comment]
@dom96 @Araq we need to update `csources` since new Magic is introduces in `lib/core/macros`.

**Comment by Araq** [comment]
No we don't.

**Pull request merged**